src/byte_sniffer.rs

```diff
@@ -0,0 +1,138 @@
+use futures_util::io::AsyncRead;
+use pin_project::pin_project;
+use std::{
+    pin::Pin,
+    sync::{
+        atomic::{AtomicBool, Ordering},
+        Arc,
+    },
+    task::{Context, Poll},
+};
+
+#[pin_project]
+pub(crate) struct ByteSniffer<Reader>
+where
+    Reader: AsyncRead,
+{
+    /// The wrapped [`AsyncRead`]er
+    #[pin]
+    inner: Reader,
+
+    /// Current length accumulator
+    current_length: usize,
+
+    /// The configured maximum length
+    max_length: usize,
+
+    upload_clamped: Option<Arc<AtomicBool>>,
+}
+
+impl<Reader> ByteSniffer<Reader>
+where
+    Reader: AsyncRead,
+{
+    pub fn new(max_length: usize, inner: Reader) -> Self {
+        let current_length = 0;
+
+        Self {
+            inner,
+            current_length,
+            max_length,
+            upload_clamped: None,
+        }
+    }
+
+    pub fn with_callback(mut self, flag: Arc<AtomicBool>) -> Self {
+        self.upload_clamped = Some(flag);
+        self
+    }
+}
+
+impl<Reader> AsyncRead for ByteSniffer<Reader>
+where
+    Reader: AsyncRead,
+{
+    fn poll_read(
+        self: Pin<&mut Self>,
+        cx: &mut Context,
+        buf: &mut [u8],
+    ) -> Poll<futures_util::io::Result<usize>> {
+        let this = self.project();
+
+        let result = this.inner.poll_read(cx, buf);
+
+        match result {
+            Poll::Ready(Ok(bytes)) => handle_ok(
+                this.current_length,
+                *this.max_length,
+                bytes,
+                this.upload_clamped,
+            ),
+            x => x,
+        }
+    }
+}
+
+fn handle_ok(
+    current_length: &mut usize,
+    max_length: usize,
+    bytes: usize,
+    upload_clamped: &Option<Arc<AtomicBool>>,
+) -> Poll<Result<usize, futures_util::io::Error>> {
+    *current_length += bytes;
+
+    Poll::Ready(match check_under_maximum(*current_length, max_length) {
+        Ok(()) => Ok(bytes),
+        Err(e) => {
+            if let Some(b) = upload_clamped {
+                b.store(true, Ordering::Relaxed)
+            }
+            Err(e.into())
+        }
+    })
+}
+
+fn check_under_maximum(current_length: usize, max_length: usize) -> Result<(), Error> {
+    if current_length > max_length {
+        Err(Error::new(max_length))
```

Comment:
I think the use of the Atomic sneaks in here - because this relies on future_util::io::Error, there is no futures io ErrorKind type that indicates "Request Too Large" in such a specific way. Perhaps ErrorKind::InvalidInput is the closest - but if there was a "unique error" type here that could indicate this, then the atomic bool would not be needed because then you could match later on in the process to generate the correct tide error instead of needing to set an operation flag here. 

Regardless, since this atomic is uncontended and used with relaxed, it's unlikely to generate a significant performance penalty. You could however consider a https://docs.rs/async-std/1.8.0/async_std/channel/fn.bounded.html bounded to size one for sending the state to async handle (but it does use more atomics under the hood).

Reply:
That all makes sense.

I didn't want to rely on checking the error type in case that was transformed or handled somewhere else in the pipeline.

You can see in the other pull request I changed it from an atomicbool to a generic callback, although I certainly considered using a one shot channel instead. I normally try to avoid callbacks if I can, because they always feel a little backwards and inside out, but I felt the ergonomics weren't too bad here